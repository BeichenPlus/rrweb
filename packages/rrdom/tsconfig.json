{
  "compilerOptions": {
    "rootDir": "src",
    "composite": true,
    "target": "ES6",
    "module": "commonjs",
    "noImplicitAny": true,
    "strictNullChecks": true,
    "removeComments": true,
    "preserveConstEnums": true,
    "sourceMap": true,
    "outDir": "build",
    "lib": ["es6", "dom"],
    "skipLibCheck": true,
    "declaration": true,
    "importsNotUsedAsValues": "error",
    "esModuleInterop": true,
    "strictBindCallApply": true
  },
  "references": [
    {
      "path": "../rrweb-snapshot"
    },
    {
      "path": "../types"
    }
  ],
  "compileOnSave": true,
  "exclude": ["test"],
  "include": [
    "src",
    "../rrweb/src/record/workers/workers.d.ts",
    "../rrweb/src/record/constructable-stylesheets.d.ts"
  ]
}
