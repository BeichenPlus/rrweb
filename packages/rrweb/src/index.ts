import record from './record';
import { Replayer } from './replay';
import { SyncReplayer } from './replay/sync-replayer';
import { _mirror } from './utils';
import * as utils from './utils';

export {
  EventType,
  IncrementalSource,
  MouseInteractions,
  ReplayerEvents,
} from '@rrweb/types';

export type { recordOptions } from './types';

const { addCustomEvent } = record;
const { freezePage } = record;

export {
  record,
  addCustomEvent,
  freezePage,
  Replayer,
  SyncReplayer,
  _mirror as mirror,
  utils,
};
