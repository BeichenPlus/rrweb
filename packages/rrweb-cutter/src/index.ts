import type {
  addedNodeMutation,
  eventWithTime,
  mousePosition,
} from '@rrweb/types';
import { IncrementalSource } from 'rrweb';
import { EventType } from 'rrweb';
import { SyncReplayer } from 'rrweb/sync-replayer';
import snapshot from './snapshot';
import { serializedNodeWithId } from 'rrweb-snapshot';
type CutterConfig = {
  points: number[];
};

export function sessionCut(
  events: eventWithTime[],
  config: CutterConfig,
): eventWithTime[][] {
  // Events length is too short so that cutting process is not needed.
  if (events.length < 2) return [events];
  const { points } = config;
  if (!points || points.length == 0) return [events];

  events = events.sort((a1, a2) => a1.timestamp - a2.timestamp);
  const totalTime = events[events.length - 1].timestamp - events[0].timestamp;

  const validSortedPoints = getValidSortedPoints(points, totalTime);
  if (validSortedPoints.length < 1) return [events];
  const results: eventWithTime[][] = [];
  const replayer = new SyncReplayer(events);
  let cutPointIndex = 0;
  const baseTime = events[0].timestamp;
  const validSortedTimestamp = validSortedPoints.map(
    (point) => baseTime + point,
  );
  replayer.play(({ index, event }) => {
    if (
      event.timestamp <= validSortedTimestamp[cutPointIndex] &&
      index + 1 < events.length
    ) {
      const nextEvent = events[index + 1];
      while (
        cutPointIndex < validSortedTimestamp.length &&
        nextEvent.timestamp > validSortedTimestamp[cutPointIndex]
      ) {
        if (results.length === 0) {
          results.push(events.slice(0, index + 1));
        }
        cutPointIndex++;
        const nextCutTimestamp =
          cutPointIndex < validSortedPoints.length
            ? validSortedTimestamp[cutPointIndex]
            : events[events.length - 1].timestamp;
        const result = cutEvents(
          events.slice(index + 1),
          replayer,
          event.timestamp,
          nextCutTimestamp,
        );
        results.push(result);
      }
      return cutPointIndex < validSortedTimestamp.length;
    }
    return false;
  });
  return results;
}

function cutEvents(
  events: eventWithTime[],
  replayer: SyncReplayer,
  currentTimestamp: number,
  endTimestamp: number,
) {
  const result: eventWithTime[] = [];
  if (replayer.latestMetaEvent) {
    const metaEvent = replayer.latestMetaEvent;
    metaEvent.timestamp = currentTimestamp;
    result.push(metaEvent);
  }
  result.push(
    ...replayer.unhandledEvents.map((e) => {
      e.timestamp = currentTimestamp + 10;
      return e;
    }),
  );
  const fullSnapshot = snapshot(replayer.virtualDom, {
    mirror: replayer.getMirror(),
  });
  if (fullSnapshot)
    result.push({
      type: EventType.FullSnapshot,
      data: {
        node: fullSnapshot,
        initialOffset: {
          top: 0,
          left: 0,
        },
      },
      timestamp: currentTimestamp,
    });
  result.push(...events.filter((event) => event.timestamp <= endTimestamp));
  return result;
}

export function pruneBranches(
  events: eventWithTime[],
  { keep }: { keep: number[] },
): eventWithTime[] {
  const result: eventWithTime[] = [];
  const replayer = new SyncReplayer(events);
  const treeSet = new Set<number>(keep);
  replayer.reversePlay(({ event }) => {
    if (event.type === EventType.FullSnapshot) {
      const { node } = event.data;
      const tree = getTreeForId(treeSet, node, keep);
      tree.forEach((id) => treeSet.add(id));
    } else if (event.type === EventType.IncrementalSnapshot) {
      if (event.data.source === IncrementalSource.Mutation) {
        const { adds, removes } = event.data;
        removes.forEach((remove) => {
          if (treeSet.has(remove.id)) treeSet.add(remove.parentId);
        });
        adds.forEach((add) => {
          const tree = getTreeForId(treeSet, add.node, keep);
          if (tree.size) {
            treeSet.add(add.parentId);
            tree.forEach((id) => treeSet.add(id));
          } else if (
            'childNodes' in add.node &&
            add.node.childNodes.length > 0
          ) {
            const tree = getTreeForId(treeSet, add.node, keep);
            if (tree.size) treeSet.add(add.parentId);
            tree.forEach((id) => treeSet.add(id));
          }
        });
      }
    }
    return true;
  });

  replayer.play(({ event }) => {
    if (
      [EventType.Meta, EventType.Load, EventType.DomContentLoaded].includes(
        event.type,
      )
    ) {
      result.push(event);
    } else if (event.type === EventType.FullSnapshot) {
      const { node } = event.data;
      const prunedNode = reconstructTreeWithIds(node, treeSet);
      if (prunedNode)
        result.push({
          ...event,
          data: {
            ...event.data,
            node: prunedNode,
          },
        } as eventWithTime);
    } else if (event.type === EventType.IncrementalSnapshot) {
      if ('positions' in event.data) {
        const { positions } = event.data;
        const prunedPositions: mousePosition[] = positions.filter((p) =>
          treeSet.has(p.id),
        );
        if (prunedPositions.length > 0)
          result.push({
            ...event,
            data: {
              ...event.data,
              positions: prunedPositions,
            },
          } as eventWithTime);
      } else if ('id' in event.data) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        if (treeSet.has(event.data.id!)) result.push(event);
      } else if (event.data.source === IncrementalSource.Mutation) {
        const { removes, adds, texts, attributes } = event.data;
        const prunedRemoves = removes.filter((remove) =>
          treeSet.has(remove.id),
        );
        const prunedAdds = adds
          .map((add) =>
            treeSet.has(add.parentId) && keep.includes(add.parentId)
              ? add
              : {
                  ...add,
                  node: reconstructTreeWithIds(add.node, treeSet),
                },
          )
          .filter((add) => Boolean(add.node)) as addedNodeMutation[];
        const prunedTexts = texts.filter((text) => treeSet.has(text.id));
        const prunedAttributes = attributes.filter((attr) =>
          treeSet.has(attr.id),
        );
        if (
          prunedRemoves.length > 0 ||
          prunedAdds.length > 0 ||
          prunedTexts.length > 0 ||
          prunedAttributes.length > 0
        )
          result.push({
            ...event,
            data: {
              ...event.data,
              removes: prunedRemoves,
              adds: prunedAdds,
              texts: prunedTexts,
              attributes: prunedAttributes,
            },
          } as eventWithTime);
      }
    }
    return true;
  });
  return result;
}

export function getTreeForId(
  treeSet: Set<number>,
  node: serializedNodeWithId,
  keepIds: number[],
): Set<number> {
  const results = new Set<number>();
  if (treeSet.has(node.id)) {
    getIdsInNode(node, keepIds).forEach((id) => results.add(id));
  } else if ('childNodes' in node) {
    for (let i = 0; i < node.childNodes.length; i++) {
      const child = node.childNodes[i];
      const childTree = getTreeForId(treeSet, child, keepIds);
      if (childTree.size > 0) {
        results.add(node.id);
        childTree.forEach((id) => results.add(id));
      }
    }
  }
  return results;
}

export function getIdsInNode(
  node: serializedNodeWithId,
  keepIds: number[],
): Array<number> {
  const results: number[] = [];
  results.push(node.id);
  if (keepIds.includes(node.id) && 'childNodes' in node) {
    for (let i = 0; i < node.childNodes.length; i++) {
      const child = node.childNodes[i];
      results.push(...getIdsInNode(child, keepIds));
    }
  }
  return results;
}

export function reconstructTreeWithIds(
  node: serializedNodeWithId,
  ids: Set<number>,
): serializedNodeWithId | undefined {
  if (ids.has(node.id)) {
    if ('childNodes' in node) {
      node.childNodes = node.childNodes
        .map((child) => reconstructTreeWithIds(child, ids))
        .filter(Boolean) as serializedNodeWithId[];
    }
    return node;
  }
  return undefined;
}

export function getValidSortedPoints(points: number[], totalTime: number) {
  const validSortedPoints = [];
  for (let i = 0; i < points.length; i++) {
    const point = points[i];
    if (point <= 0 || point >= totalTime) continue;
    validSortedPoints.push(point);
  }
  return validSortedPoints.sort();
}
