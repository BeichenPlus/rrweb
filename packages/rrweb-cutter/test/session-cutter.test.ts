/**
 * @jest-environment jsdom
 */
import path from 'path';
import fs from 'fs';
import { createMirror, snapshot, NodeType } from 'rrweb-snapshot';
import { EventType } from 'rrweb';
import { SyncReplayer } from 'rrweb';
import type { eventWithTime } from '@rrweb/types';
import { RRDocument, buildFromDom, printRRDom } from 'rrdom';
import { sessionCut, getValidSortedPoints, pruneBranches } from '../src';
import { snapshot as RRDomSnapshot } from '../src/snapshot';
import { events as mutationEvents } from './events/mutation.event';
import { eventsFn as inlineStyleEvents } from './events/inline-style.event';
import { assertSnapshot } from 'rrweb/test/utils';

describe('session cutter', () => {
  it('should return the same events if the events length is too short', () => {
    const events1: eventWithTime[] = [];
    const config = { points: [10] };
    expect(sessionCut(events1, config)).toEqual([events1]);

    const events2: eventWithTime[] = [
      {
        type: EventType.Load,
        data: {},
        timestamp: 1,
      } as eventWithTime,
    ];
    expect(sessionCut(events2, config)).toEqual([events2]);
  });

  it('should return the same events if the points length is 0', () => {
    const events: eventWithTime[] = [
      {
        type: EventType.Load,
        data: {},
        timestamp: 1,
      } as eventWithTime,
      {
        type: EventType.Meta,
        data: {},
        timestamp: 2,
      } as eventWithTime,
    ];
    const config = { points: [] };
    expect(sessionCut(events, config)).toEqual([events]);
  });

  it('should sort and validate cutting points array', () => {
    const inputPoints = [10, 250.5, -10, -1, 0, 100];
    expect(getValidSortedPoints([], 100)).toEqual([]);
    expect(getValidSortedPoints(inputPoints, 11)).toEqual([10]);
    expect(getValidSortedPoints(inputPoints, 10)).toEqual([]);
    expect(getValidSortedPoints(inputPoints, 100)).toEqual([10]);
    expect(getValidSortedPoints(inputPoints, 300)).toEqual([10, 100, 250.5]);
  });

  describe('Build full snapshot events from RRDom', () => {
    it("should build full snapshot events from RRDom's mirror: main.html", () => {
      document.write(getHtml('main.html'));
      const rrdom = new RRDocument();
      const mirror = createMirror();
      // the full snapshot that is built on jsdom
      const originalSnapshot = snapshot(document, { mirror });

      // Create a RRDom according to the jsdom (real dom).
      buildFromDom(document, mirror, rrdom);

      const newFullSnapshot = RRDomSnapshot(rrdom, {
        mirror: rrdom.mirror,
      });
      expect(newFullSnapshot).toEqual(originalSnapshot);
    });
  });

  describe('Cut the session events from several time points', () => {
    it('should cut the simplest mutation events', () => {
      const events = mutationEvents as eventWithTime[];
      const result = sessionCut(events, { points: [1000, 2000] });
      expect(result).toHaveLength(3);

      // all events before 1000ms
      const sessionBefore1s = result[0];
      const cutPoint1Length = 5;
      expect(sessionBefore1s).toHaveLength(cutPoint1Length);
      // These events are directly sliced from the original events.
      expect(sessionBefore1s).toEqual(events.slice(0, cutPoint1Length));

      // all events between 1000ms and 2000ms
      const sessionBetween1s2s = result[1];
      expect(sessionBetween1s2s).toHaveLength(3);
      expect(sessionBetween1s2s[0].type).toEqual(EventType.Meta);
      expect(sessionBetween1s2s[1].type).toEqual(EventType.FullSnapshot);
      expect(sessionBetween1s2s[2].type).toEqual(EventType.IncrementalSnapshot);
      let replayer = new SyncReplayer(sessionBetween1s2s.slice(0, 2)); // only play meta and full snapshot events
      replayer.play();
      // screenshot at 1000ms
      expect(
        printRRDom(replayer.virtualDom, replayer.getMirror()),
      ).toMatchSnapshot('screenshot at 1000ms');

      // all events after 2000ms
      const sessionAfter2s = result[2];
      expect(sessionAfter2s).toHaveLength(3);
      expect(sessionAfter2s[0].type).toEqual(EventType.Meta);
      expect(sessionAfter2s[1].type).toEqual(EventType.FullSnapshot);
      expect(sessionAfter2s[2].type).toEqual(EventType.IncrementalSnapshot);
      replayer = new SyncReplayer(sessionAfter2s);
      replayer.play(({ index }) => {
        if (index === 1)
          // full snapshot
          // screen shot at 2000ms
          expect(
            printRRDom(replayer.virtualDom, replayer.getMirror()),
          ).toMatchSnapshot('screenshot at 2000ms');
      });
      // screen shot at 3000ms
      expect(
        printRRDom(replayer.virtualDom, replayer.getMirror()),
      ).toMatchSnapshot('screenshot at 3000ms');
    });
  });

  it('should cut events with inline styles', () => {
    const events = inlineStyleEvents() as eventWithTime[];
    const result = sessionCut(events, { points: [1000] });
    expect(result).toHaveLength(2);
    // all events before 1000ms
    const sessionBefore1s = result[0];
    const cutPoint1Length = 5;
    expect(sessionBefore1s).toHaveLength(cutPoint1Length);
    // These events are directly sliced from the original events.
    expect(sessionBefore1s).toEqual(events.slice(0, cutPoint1Length));

    // all events after 1000ms
    const sessionAfter1s = result[1];
    expect(sessionAfter1s).toHaveLength(3);
    const replayer = new SyncReplayer(sessionAfter1s.slice(0, 2)); // only play meta and full snapshot events
    replayer.play();
    // screenshot at 1000ms
    expect(
      printRRDom(replayer.virtualDom, replayer.getMirror()),
    ).toMatchSnapshot('screenshot at 1000ms');
  });
});

describe('pruneBranches', () => {
  it("should cut branches that doesn't include id", () => {
    const events = inlineStyleEvents() as eventWithTime[];
    const result = pruneBranches(events, { keep: [14] });
    expect(result).toHaveLength(5);
    const replayer = new SyncReplayer(result);
    replayer.play();

    expect(
      printRRDom(replayer.virtualDom, replayer.getMirror()),
    ).toMatchSnapshot('pruned all but 14');
  });

  it("should remove mutations that don't include ids", () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [
          {
            id: 15,
            value: 'Kept',
          },
          {
            id: 1001,
            value: 'Cut',
          },
        ],
        attributes: [
          {
            id: 14,
            attributes: {
              'data-attr': 'Kept',
            },
          },
          {
            id: 1002,
            attributes: {
              'data-attr': 'Cut',
            },
          },
        ],
        removes: [
          {
            parentId: 14,
            id: 15,
          },
          {
            parentId: 1002,
            id: 1003,
          },
        ],
        adds: [
          {
            parentId: 14,
            nextId: null,
            node: {},
          },
          {
            parentId: 1001,
            nextId: null,
            node: {},
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [14] });
    expect(result[result.length - 1]).toMatchSnapshot();
  });

  it('should remove branches based on nodes that came in after fullsnapshot', () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [],
        attributes: [],
        removes: [],
        adds: [
          {
            parentId: 14,
            nextId: null,
            node: {
              id: 99,
              type: NodeType.Element,
              tagName: 'canvas',
              attributes: {},
              childNodes: [],
            },
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [99] });
    assertSnapshot(result);
  });
  it('should remove branches based on child nodes that came in after fullsnapshot', () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [],
        attributes: [],
        removes: [],
        adds: [
          {
            parentId: 14,
            nextId: null,
            node: {
              id: 98,
              type: NodeType.Element,
              tagName: 'main',
              attributes: {},
              childNodes: [
                {
                  id: 99,
                  type: NodeType.Element,
                  tagName: 'canvas',
                  attributes: {},
                  childNodes: [],
                },
              ],
            },
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [99] });
    assertSnapshot(result);
  });

  it('should keep branches where target child nodes was, and gets moved to', () => {
    const mutationEvent = {
      type: EventType.IncrementalSnapshot,
      data: {
        source: 0,
        texts: [],
        attributes: [],
        removes: [
          {
            parentId: 14,
            id: 15,
          },
        ],
        adds: [
          {
            parentId: 5,
            nextId: null,
            node: { type: 3, textContent: '      \n    ', id: 15 },
          },
        ],
      },
    };
    const events = [...inlineStyleEvents(), mutationEvent] as eventWithTime[];
    const result = pruneBranches(events, { keep: [15] });
    assertSnapshot(result);
  });
});

function getHtml(fileName: string) {
  const filePath = path.resolve(__dirname, `./html/${fileName}`);
  return fs.readFileSync(filePath, 'utf8');
}
