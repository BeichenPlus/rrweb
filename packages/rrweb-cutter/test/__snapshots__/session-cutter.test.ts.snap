// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruneBranches should cut branches that doesn't include id: pruned all but 14 1`] = `
"1 RRDocument
  -1 RRDocumentType
  2 HTML 
    12 BODY 
      14 DIV id="root" 
        15 RRText text="      \\n    "
"
`;

exports[`pruneBranches should keep branches where target child nodes was, and gets moved to 1`] = `
"[
  {
    "type": 0,
    "data": {}
  },
  {
    "type": 1,
    "data": {}
  },
  {
    "type": 4,
    "data": {
      "href": "about:blank",
      "width": 1512,
      "height": 395
    }
  },
  {
    "type": 2,
    "data": {
      "node": {
        "type": 0,
        "childNodes": [
          {
            "type": 2,
            "tagName": "html",
            "attributes": {},
            "childNodes": [
              {
                "type": 2,
                "tagName": "head",
                "attributes": {},
                "childNodes": [
                  {
                    "type": 2,
                    "tagName": "meta",
                    "attributes": {
                      "charset": "utf-8"
                    },
                    "childNodes": [],
                    "id": 5
                  }
                ],
                "id": 3
              },
              {
                "type": 2,
                "tagName": "body",
                "childNodes": [
                  {
                    "type": 2,
                    "tagName": "div",
                    "attributes": {
                      "id": "root"
                    },
                    "childNodes": [
                      {
                        "type": 3,
                        "textContent": "      \\n    ",
                        "id": 15
                      }
                    ],
                    "id": 14
                  }
                ],
                "id": 12
              }
            ],
            "id": 2
          }
        ],
        "compatMode": "BackCompat",
        "id": 1
      },
      "initialOffset": {
        "left": 0,
        "top": 0
      }
    }
  },
  {
    "type": 3,
    "data": {
      "source": 2,
      "type": 2,
      "id": 15
    }
  },
  {
    "type": 3,
    "data": {
      "source": 0,
      "texts": [],
      "attributes": [],
      "removes": [
        {
          "parentId": 14,
          "id": 15
        }
      ],
      "adds": [
        {
          "parentId": 5,
          "nextId": null,
          "node": {
            "type": 3,
            "textContent": "      \\n    ",
            "id": 15
          }
        }
      ]
    }
  }
]"
`;

exports[`pruneBranches should remove branches based on child nodes that came in after fullsnapshot 1`] = `
"[
  {
    "type": 0,
    "data": {}
  },
  {
    "type": 1,
    "data": {}
  },
  {
    "type": 4,
    "data": {
      "href": "about:blank",
      "width": 1512,
      "height": 395
    }
  },
  {
    "type": 2,
    "data": {
      "node": {
        "type": 0,
        "childNodes": [
          {
            "type": 2,
            "tagName": "html",
            "attributes": {},
            "childNodes": [
              {
                "type": 2,
                "tagName": "body",
                "childNodes": [
                  {
                    "type": 2,
                    "tagName": "div",
                    "attributes": {
                      "id": "root"
                    },
                    "childNodes": [],
                    "id": 14
                  }
                ],
                "id": 12
              }
            ],
            "id": 2
          }
        ],
        "compatMode": "BackCompat",
        "id": 1
      },
      "initialOffset": {
        "left": 0,
        "top": 0
      }
    }
  },
  {
    "type": 3,
    "data": {
      "source": 0,
      "texts": [],
      "attributes": [],
      "removes": [],
      "adds": [
        {
          "parentId": 14,
          "nextId": null,
          "node": {
            "id": 98,
            "type": 2,
            "tagName": "main",
            "attributes": {},
            "childNodes": [
              {
                "id": 99,
                "type": 2,
                "tagName": "canvas",
                "attributes": {},
                "childNodes": []
              }
            ]
          }
        }
      ]
    }
  }
]"
`;

exports[`pruneBranches should remove branches based on nodes that came in after fullsnapshot 1`] = `
"[
  {
    "type": 0,
    "data": {}
  },
  {
    "type": 1,
    "data": {}
  },
  {
    "type": 4,
    "data": {
      "href": "about:blank",
      "width": 1512,
      "height": 395
    }
  },
  {
    "type": 2,
    "data": {
      "node": {
        "type": 0,
        "childNodes": [
          {
            "type": 2,
            "tagName": "html",
            "attributes": {},
            "childNodes": [
              {
                "type": 2,
                "tagName": "body",
                "childNodes": [
                  {
                    "type": 2,
                    "tagName": "div",
                    "attributes": {
                      "id": "root"
                    },
                    "childNodes": [],
                    "id": 14
                  }
                ],
                "id": 12
              }
            ],
            "id": 2
          }
        ],
        "compatMode": "BackCompat",
        "id": 1
      },
      "initialOffset": {
        "left": 0,
        "top": 0
      }
    }
  },
  {
    "type": 3,
    "data": {
      "source": 0,
      "texts": [],
      "attributes": [],
      "removes": [],
      "adds": [
        {
          "parentId": 14,
          "nextId": null,
          "node": {
            "id": 99,
            "type": 2,
            "tagName": "canvas",
            "attributes": {},
            "childNodes": []
          }
        }
      ]
    }
  }
]"
`;

exports[`pruneBranches should remove mutations that don't include ids 1`] = `
{
  "data": {
    "adds": [
      {
        "nextId": null,
        "node": {},
        "parentId": 14,
      },
    ],
    "attributes": [
      {
        "attributes": {
          "data-attr": "Kept",
        },
        "id": 14,
      },
    ],
    "removes": [
      {
        "id": 15,
        "parentId": 14,
      },
    ],
    "source": 0,
    "texts": [
      {
        "id": 15,
        "value": "Kept",
      },
    ],
  },
  "type": 3,
}
`;

exports[`session cutter Cut the session events from several time points should cut the simplest mutation events: screenshot at 1000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 META charset="utf-8" 
      5 RRText text="    \\n  "
    6 BODY 
      16 RRText text="\\n    "
      17 DIV id="container" 
        34 DIV 
          35 RRText text="1"
        33 DIV 
          36 RRText text="2"
        32 DIV 
          37 RRText text="3"
        31 DIV 
          38 RRText text="4"
        29 DIV 
          30 RRText text="5"
      19 RRText text="\\n  \\n  "
"
`;

exports[`session cutter Cut the session events from several time points should cut the simplest mutation events: screenshot at 2000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 META charset="utf-8" 
      5 RRText text="    \\n  "
    6 BODY 
      16 RRText text="\\n    "
      17 DIV id="container" 
        29 DIV 
          30 RRText text="5"
        31 DIV 
          38 RRText text="4"
        32 DIV 
          37 RRText text="3"
        33 DIV 
          36 RRText text="2"
        34 DIV 
          35 RRText text="1"
      19 RRText text="\\n  \\n  "
"
`;

exports[`session cutter Cut the session events from several time points should cut the simplest mutation events: screenshot at 3000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 META charset="utf-8" 
      5 RRText text="    \\n  "
    6 BODY 
      16 RRText text="\\n    "
      17 DIV id="container" 
      19 RRText text="\\n  \\n  "
"
`;

exports[`session cutter should cut events with inline styles: screenshot at 1000ms 1`] = `
"1 RRDocument
  2 HTML 
    3 HEAD 
      4 RRText text="\\n    "
      5 META charset="utf-8" 
      6 RRText text="\\n    "
      7 STYLE 
        -1 RRText text="#root { background: yellow; width: 10px; height: 10px; }"
      8 RRText text="\\n    "
      9 STYLE 
        -1 RRText text=".block { width: 20px; height: 20px; background: red; }"
      10 RRText text="\\n  "
    11 RRText text="\\n  "
    12 BODY 
      13 RRText text="\\n    "
      14 DIV id="root" 
        15 RRText text="      \\n    "
      16 RRText text="\\n    "
      17 DIV class="block" 
"
`;
